graphql_compiler/tests/test_compiler.py

```diff
@@ -9268,75 +9315,91 @@ def test_simple_optional_recurse(self):
+                    [Animal_2].parent AS [Animal_in_Animal_ParentOf__parent],
+                    [Animal_2].uuid AS [Animal_in_Animal_ParentOf__uuid]
+                FROM
+                    db_1.schema_1.[Animal] AS [Animal_1]
+                    LEFT OUTER JOIN db_1.schema_1.[Animal] AS [Animal_2]
+                        ON [Animal_1].parent = [Animal_2].uuid
+            ),
+            anon_2(name, parent, uuid, __cte_key, __cte_depth) AS (
+                SELECT
+                    anon_1.[Animal_in_Animal_ParentOf__name] AS name,
+                    anon_1.[Animal_in_Animal_ParentOf__parent] AS parent,
+                    anon_1.[Animal_in_Animal_ParentOf__uuid] AS uuid,
+                    anon_1.[Animal_in_Animal_ParentOf__uuid] AS __cte_key,
+                    0 AS __cte_depth
+                FROM
+                    anon_1
+                UNION ALL
                 SELECT
                     [Animal_3].name AS name,
                     [Animal_3].parent AS parent,
                     [Animal_3].uuid AS uuid,
-                    [Animal_3].uuid AS __cte_key,
-                    0 AS __cte_depth
+                    anon_2.__cte_key AS __cte_key,
+                    anon_2.__cte_depth + 1 AS __cte_depth
                 FROM
-                    db_1.schema_1.[Animal] AS [Animal_3]
-                UNION ALL
-                    SELECT
-                        [Animal_4].name AS name,
-                        [Animal_4].parent AS parent,
-                        [Animal_4].uuid AS uuid,
-                        anon_1.__cte_key AS __cte_key,
-                        anon_1.__cte_depth + 1 AS __cte_depth
-                    FROM
-                        anon_1
-                        JOIN db_1.schema_1.[Animal] AS [Animal_4]
-                            ON anon_1.uuid = [Animal_4].parent
-                    WHERE anon_1.__cte_depth < 3
+                    anon_2
+                    JOIN db_1.schema_1.[Animal] AS [Animal_3]
+                        ON anon_2.uuid = [Animal_3].parent
+                    WHERE anon_2.__cte_depth < 3
             )
             SELECT
-                [Animal_1].name AS child_name,
-                [Animal_2].name AS name,
-                anon_1.name AS self_and_ancestor_name
-            FROM
-                db_1.schema_1.[Animal] AS [Animal_2]
-                LEFT OUTER JOIN db_1.schema_1.[Animal] AS [Animal_1]
-                    ON [Animal_2].parent = [Animal_1].uuid
-                LEFT OUTER JOIN anon_1 ON [Animal_1].uuid = anon_1.__cte_key
-            WHERE
-                anon_1.__cte_key IS NOT NULL OR [Animal_1].uuid IS NULL
+                anon_1.[Animal_in_Animal_ParentOf__name] AS child_name,
+                anon_1.[Animal__name] AS name,
+                anon_2.name AS self_and_ancestor_name
+            FROM anon_1, anon_2
         """
+        # TODO Add an integration test for this query to make sure the recurse preserves
```

Comment:
Please use the `TODO(<author>):` format so we can track TODO counts and locations with standard tools.